### DISCUSSION TITLE
No Softmax in GCN example

### DISCUSSION THREAD

**Discussion by utkarsh0902311047** [pyg-team/pytorch_geometric#7726] [Q&A]
Hi,
I don't understand why there is no softmax layer in the GCN example mentioned in the original semi-supervised learning paper.
Also, please explain to me a little about where multi-layer perceptron is used in two layers of GCN.

Thank you

**Comment by akihironitta** [comment]
> I don't understand why there is no softmax layer in the GCN example mentioned in the original semi-supervised learning paper.

@utkarsh0902311047 I believe that's because the loss used in the example is `torch.nn.functional.cross_entropy` that doesn't require inputs to be normalised as documented in [the PyTorch docs](https://pytorch.org/docs/2.0/generated/torch.nn.CrossEntropyLoss.html#torch.nn.CrossEntropyLoss):

> The input is expected to contain the unnormalized logits for each class (which do not need to be positive or sum to 1, in general).

**Comment by utkarsh0902311047** [comment]
@akihironitta 
Thank you for the explanation.
Please also help with my second question, "where multi-layer perceptron is used in two layers of GCN?"

**Reply by rusty1s** [reply]
Each GNN layer performs a linear transformation, so stacking these layers + non-linearities will resemble MLPs (in case of empty edge indices).
